### DISCUSSION TITLE
Custom JSON (de)serialisation not working on POST

### DISCUSSION THREAD

**Discussion by charlielito** [fastapi/fastapi#8972] [Questions]
Hi there!
I'm trying to send a post with a JSON that might contain a `\n` character in one of its fields. I tried what is explained here https://pydantic-docs.helpmanual.io/usage/exporting_models/ with no success. I can run the example there with no problems but when used in the app it crashes when it tries to parse the input json.

The sample code:

```Python
import json
from fastapi import FastAPI
from pydantic import BaseModel

def new_json_loads(*args, **kwargs):
    kwargs.update(dict(strict=False))
    return json.loads(*args, **kwargs)

class Request(BaseModel):
    name = "John Doe"
    class Config:
        json_loads = new_json_loads

r = Request.parse_raw('{\n  "name": "john \n doe"\n}')
# It runs!
print(r)

app = FastAPI()

# It does not run here when a request arrives
@app.post("/test")
def read_root(request: Request):
    return request
```
### Description

* Open the browser and call the endpoint http://0.0.0.0:8000/docs#/default/read_root_test_post.
* `Try it out` with: `{
  "name": "john 
doe"
}`. Don't forget to enter a new line after the word `John`
* The 422 error returns
```
{
  "detail": [
    {
      "loc": [
        "body",
        18
      ],
      "msg": "Invalid control character at: line 2 column 17 (char 18)",
      "type": "value_error.jsondecode",
      "ctx": {
        "msg": "Invalid control character at",
        "doc": "{\n  \"name\": \"john \ndoe\"\n}",
        "pos": 18,
        "lineno": 2,
        "colno": 17
      }
    }
  ]
}
```
### Environment

* OS: [e.g. Linux / Windows / macOS]: Linux 20.04
* FastAPI Version [e.g. 0.3.0]: 0.61.1
* Python version: 3.7.8

Thanks for your help!

**Comment by oryba** [comment]
Hi @charlielito! 

As you can see **[here](https://github.com/tiangolo/fastapi/blob/9e6b069c099d0b63df6e4c754497e0dffbffc5b5/fastapi/routing.py#L171)** in sources, the custom `json_loads` won't be called at all. 

But anyway, the issue is that you're sending a malformed JSON - line breaks inside keys and values are not allowed. By pressing Enter you're creating a payload like this:
```
b'{"name":"john\n doe"}'
```

And it's invalid from the JSON standpoint. If you need to store line breaks in the value, you should escape `\` character, so the payload would be like this:
```
b'{ "name": "john\\n doe" }'
```

Just try to send `{ "name": "john\n doe" }` in the same way you've described - but without visual line breaks. Then \n would be preserved.

**Comment by charlielito** [comment]
Thanks for taking the time to answer.
I know that that way the JSON is malformed, that's why I want to add a custom `json_loads`, specifically with the parameter `strict=False` to be able to parse it. I want this because I am using the swagger ui to send a sql query in a JSON field, and such query includes break lines when pasting the query from another place. 

The question is if it is even possible that the custom `json_loads` is used?

**Comment by FionnD** [comment]
Hi 👋 

This is an issue for me as well, when a user sends a load that contains an accidental `\` I would like an easy way to remove that and process the request without just returning an error message. I see a very long dicussion about a similar  issue to this https://github.com/samuelcolvin/pydantic/issues/1098 on pydantic site.

**Comment by NixonInnes** [comment]
I'm also looking for a way to use a custom `json_loads`
My use case, is my data model contains a `pandas.DataFrame` which I encode using a custom encoder, and would like to unzip it at the other end.

```python
def ts_json_loads(v):
    dic = json.loads(v)
    dic["data"] = pd.read_json(unzip_str(dic["data"]))
    return dic

class Timeseries(BaseModel):
    name: str
    data: pd.DataFrame
    id: Optional[int] = None
    start: Optional[datetime] = None
    end: Optional[datetime] = None

    class Config:
        arbitrary_types_allowed = True
        json_encoders = {
            pd.DataFrame: lambda df: zip_str(df.to_json())
        }
        json_loads = ts_json_loads

def zip_str(s):
    return base64.b64encode(
        zlib.compress(
            s.encode("utf-8")
        )
    ).decode("ascii")

def unzip_str(s):
    try:
        return zlib.decompress(base64.b64decode(s)).decode("utf-8")
    except:
        raise RuntimeError("Could not decode/unzip the contents")
```

**Comment by xaviml** [comment]
Having same problem. I would like to throw an error when a duplicated key is present, and this is possible through pydantic, but FastAPI is not validating the JSON loader with pydantic, it is doing it with Starlette.
